File: dpp/lexer.py
"""Lexer for the D++ language."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto

from .errors import LexerError


class TokenType(Enum):
    LEFT_PAREN = auto()
    RIGHT_PAREN = auto()
    LEFT_BRACE = auto()
    RIGHT_BRACE = auto()
    COMMA = auto()
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    ASSIGN = auto()
    EQEQ = auto()
    BANGEQ = auto()
    GT = auto()
    GTE = auto()
    LT = auto()
    LTE = auto()
    RANGE = auto()
    IDENTIFIER = auto()
    NUMBER = auto()
    STRING = auto()
    LET = auto()
    FN = auto()
    RETURN = auto()
    IF = auto()
    ELSE = auto()
    FOR = auto()
    IN = auto()
    PRINT = auto()
    TRUE = auto()
    FALSE = auto()
    NULL = auto()
    EOF = auto()


KEYWORDS = {
    "let": TokenType.LET,
    "fn": TokenType.FN,
    "return": TokenType.RETURN,
    "if": TokenType.IF,
    "else": TokenType.ELSE,
    "for": TokenType.FOR,
    "in": TokenType.IN,
    "print": TokenType.PRINT,
    "true": TokenType.TRUE,
    "false": TokenType.FALSE,
    "null": TokenType.NULL,
}


@dataclass(slots=True)
class Token:
    type: TokenType
    lexeme: str
    literal: object
    line: int
    column: int


class Lexer:
    def __init__(self, source: str) -> None:
        self.source = source
        self.length = len(source)
        self.start = 0
        self.current = 0
        self.line = 1
        self.column = 1
        self.start_line = 1
        self.start_column = 1
        self.tokens: list[Token] = []

    def tokenize(self) -> list[Token]:
        while not self._is_at_end():
            self.start = self.current
            self.start_line = self.line
            self.start_column = self.column
            self._scan_token()

        self.tokens.append(Token(TokenType.EOF, "", None, self.line, self.column))
        return self.tokens

    def _scan_token(self) -> None:
        char = self._advance()
        match char:
            case "(":
                self._add_token(TokenType.LEFT_PAREN)
            case ")":
                self._add_token(TokenType.RIGHT_PAREN)
            case "{":
                self._add_token(TokenType.LEFT_BRACE)
            case "}":
                self._add_token(TokenType.RIGHT_BRACE)
            case ",":
                self._add_token(TokenType.COMMA)
            case "+":
                self._add_token(TokenType.PLUS)
            case "-":
                self._add_token(TokenType.MINUS)
            case "*":
                self._add_token(TokenType.STAR)
            case "/":
                if self._match("/"):
                    while self._peek() not in ("\n", "\0"):
                        self._advance()
                else:
                    self._add_token(TokenType.SLASH)
            case "=":
                token_type = TokenType.EQEQ if self._match("=") else TokenType.ASSIGN
                self._add_token(token_type)
            case "!":
                if self._match("="):
                    self._add_token(TokenType.BANGEQ)
                else:
                    self._error("Unexpected character '!'.")
            case ">":
                token_type = TokenType.GTE if self._match("=") else TokenType.GT
                self._add_token(token_type)
            case "<":
                token_type = TokenType.LTE if self._match("=") else TokenType.LT
                self._add_token(token_type)
            case ".":
                if self._match("."):
                    self._add_token(TokenType.RANGE)
                else:
                    self._error("Unexpected character '.'.")
            case " " | "\r" | "\t":
                return
            case "\n":
                return
            case "\"":
                self._string()
            case _:
                if char.isdigit():
                    self._number()
                elif char.isalpha() or char == "_":
                    self._identifier()
                else:
                    self._error(f"Unexpected character '{char}'.")

    def _identifier(self) -> None:
        while self._peek().isalnum() or self._peek() == "_":
            self._advance()

        lexeme = self.source[self.start : self.current]
        token_type = KEYWORDS.get(lexeme, TokenType.IDENTIFIER)
        literal = None
        if token_type == TokenType.TRUE:
            literal = True
        elif token_type == TokenType.FALSE:
            literal = False
        elif token_type == TokenType.NULL:
            literal = None

        self._add_token(token_type, literal)

    def _number(self) -> None:
        while self._peek().isdigit():
            self._advance()

        if self._peek() == "." and self._peek_next().isdigit():
            self._advance()
            while self._peek().isdigit():
                self._advance()

        lexeme = self.source[self.start : self.current]
        literal = float(lexeme) if "." in lexeme else int(lexeme)
        self._add_token(TokenType.NUMBER, literal)

    def _string(self) -> None:
        value_chars: list[str] = []

        while not self._is_at_end():
            char = self._advance()
            if char == "\"":
                self._add_token(TokenType.STRING, "".join(value_chars))
                return
            if char == "\\":
                if self._is_at_end():
                    self._error("Unterminated string.")
                escape = self._advance()
                value_chars.append(self._translate_escape(escape))
                continue
            value_chars.append(char)

        self._error("Unterminated string.")

    def _translate_escape(self, escape: str) -> str:
        mapping = {
            "\"": "\"",
            "\\": "\\",
            "n": "\n",
            "t": "\t",
            "r": "\r",
        }
        if escape not in mapping:
            self._error(f"Unsupported escape sequence '\\{escape}'.")
        return mapping[escape]

    def _add_token(self, token_type: TokenType, literal: object = None) -> None:
        lexeme = self.source[self.start : self.current]
        self.tokens.append(Token(token_type, lexeme, literal, self.start_line, self.start_column))

    def _advance(self) -> str:
        char = self.source[self.current]
        self.current += 1
        if char == "\n":
            self.line += 1
            self.column = 1
        else:
            self.column += 1
        return char

    def _match(self, expected: str) -> bool:
        if self._is_at_end() or self.source[self.current] != expected:
            return False
        self._advance()
        return True

    def _peek(self) -> str:
        if self._is_at_end():
            return "\0"
        return self.source[self.current]

    def _peek_next(self) -> str:
        if self.current + 1 >= self.length:
            return "\0"
        return self.source[self.current + 1]

    def _is_at_end(self) -> bool:
        return self.current >= self.length

    def _error(self, message: str) -> None:
        raise LexerError(f"Error: {message} at line {self.start_line}, column {self.start_column}")

File: src/lexer.cpp
#include "dpp/lexer.hpp"

#include <cctype>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>

#include "dpp/error.hpp"

namespace dpp {

namespace {

const std::unordered_map<std::string, TokenType> kKeywords = {
    {"let", TokenType::Let},
    {"fn", TokenType::Fn},
    {"return", TokenType::Return},
    {"if", TokenType::If},
    {"else", TokenType::Else},
    {"while", TokenType::While},
    {"for", TokenType::For},
    {"in", TokenType::In},
    {"break", TokenType::Break},
    {"continue", TokenType::Continue},
    {"and", TokenType::And},
    {"or", TokenType::Or},
    {"print", TokenType::Print},
    {"true", TokenType::True},
    {"false", TokenType::False},
    {"null", TokenType::Null},
};

}  // namespace

Lexer::Lexer(std::string source) : source_(std::move(source)) {}

std::vector<Token> Lexer::tokenize() {
    while (!isAtEnd()) {
        start_ = current_;
        startLine_ = line_;
        startColumn_ = column_;
        scanToken();
    }

    tokens_.push_back(Token{TokenType::EndOfFile, "", std::monostate{}, line_, column_});
    return tokens_;
}

void Lexer::scanToken() {
    const char character = advance();

    switch (character) {
        case '(':
            addToken(TokenType::LeftParen);
            return;
        case ')':
            addToken(TokenType::RightParen);
            return;
        case '{':
            addToken(TokenType::LeftBrace);
            return;
        case '}':
            addToken(TokenType::RightBrace);
            return;
        case '[':
            addToken(TokenType::LeftBracket);
            return;
        case ']':
            addToken(TokenType::RightBracket);
            return;
        case ',':
            addToken(TokenType::Comma);
            return;
        case '+':
            addToken(TokenType::Plus);
            return;
        case '-':
            addToken(TokenType::Minus);
            return;
        case '*':
            addToken(TokenType::Star);
            return;
        case '/':
            if (match('/')) {
                while (peek() != '\n' && peek() != '\0') {
                    advance();
                }
            } else {
                addToken(TokenType::Slash);
            }
            return;
        case '%':
            addToken(TokenType::Percent);
            return;
        case '=':
            addToken(match('=') ? TokenType::EqualEqual : TokenType::Assign);
            return;
        case '!':
            if (match('=')) {
                addToken(TokenType::BangEqual);
                return;
            }
            addToken(TokenType::Bang);
            return;
        case '&':
            if (match('&')) {
                addToken(TokenType::And);
                return;
            }
            error("Unexpected character '&'.");
        case '|':
            if (match('|')) {
                addToken(TokenType::Or);
                return;
            }
            error("Unexpected character '|'.");
        case '>':
            addToken(match('=') ? TokenType::GreaterEqual : TokenType::Greater);
            return;
        case '<':
            addToken(match('=') ? TokenType::LessEqual : TokenType::Less);
            return;
        case '.':
            if (match('.')) {
                addToken(TokenType::Range);
                return;
            }
            error("Unexpected character '.'.");
        case ' ':
        case '\r':
        case '\t':
        case '\n':
            return;
        case '"':
            string();
            return;
        default:
            if (std::isdigit(static_cast<unsigned char>(character))) {
                number();
                return;
            }
            if (std::isalpha(static_cast<unsigned char>(character)) || character == '_') {
                identifier();
                return;
            }
            error(std::string("Unexpected character '") + character + "'.");
    }
}

void Lexer::identifier() {
    while (std::isalnum(static_cast<unsigned char>(peek())) || peek() == '_') {
        advance();
    }

    const std::string lexeme = source_.substr(start_, current_ - start_);
    const auto found = kKeywords.find(lexeme);
    if (found != kKeywords.end()) {
        addToken(found->second);
        return;
    }

    addToken(TokenType::Identifier);
}

void Lexer::number() {
    while (std::isdigit(static_cast<unsigned char>(peek()))) {
        advance();
    }

    if (peek() == '.' && peekNext() != '.' && std::isdigit(static_cast<unsigned char>(peekNext()))) {
        advance();
        while (std::isdigit(static_cast<unsigned char>(peek()))) {
            advance();
        }
    }

    const std::string lexeme = source_.substr(start_, current_ - start_);
    addToken(TokenType::Number, std::stod(lexeme));
}

void Lexer::string() {
    std::string value;

    while (!isAtEnd()) {
        const char character = advance();
        if (character == '"') {
            addToken(TokenType::String, value);
            return;
        }

        if (character == '\\') {
            if (isAtEnd()) {
                error("Unterminated string.");
            }

            const char escaped = advance();
            switch (escaped) {
                case '"':
                    value.push_back('"');
                    break;
                case '\\':
                    value.push_back('\\');
                    break;
                case 'n':
                    value.push_back('\n');
                    break;
                case 't':
                    value.push_back('\t');
                    break;
                case 'r':
                    value.push_back('\r');
                    break;
                default:
                    error(std::string("Unsupported escape sequence '\\") + escaped + "'.");
            }
            continue;
        }

        value.push_back(character);
    }

    error("Unterminated string.");
}

void Lexer::addToken(TokenType type) {
    addToken(type, std::monostate{});
}

void Lexer::addToken(TokenType type, TokenLiteral literal) {
    tokens_.push_back(Token{
        type,
        source_.substr(start_, current_ - start_),
        std::move(literal),
        startLine_,
        startColumn_,
    });
}

char Lexer::advance() {
    const char character = source_[current_++];
    if (character == '\n') {
        ++line_;
        column_ = 1;
    } else {
        ++column_;
    }
    return character;
}

bool Lexer::match(char expected) {
    if (isAtEnd() || source_[current_] != expected) {
        return false;
    }

    advance();
    return true;
}

char Lexer::peek() const {
    if (isAtEnd()) {
        return '\0';
    }
    return source_[current_];
}

char Lexer::peekNext() const {
    if (current_ + 1 >= source_.size()) {
        return '\0';
    }
    return source_[current_ + 1];
}

bool Lexer::isAtEnd() const {
    return current_ >= source_.size();
}

void Lexer::error(const std::string& message) const {
    std::ostringstream stream;
    stream << "Error: " << message << " at line " << startLine_ << ", column " << startColumn_;
    throw LexerError(stream.str());
}

}  // namespace dpp
